### PR TITLE
Initial lockfile implementation

### PR DESCRIPTION
**Pull request by edkolev**
This is an initial lockfile implementation, as discussed in #641  

I decided to add a file `scripts/lockfile/lockfile.go` which is used for testing.

### CODE REVIEW

**Comment by edkolev** [comment]
@richardartoul I think this PR is ready to be reviewed. Please let me know if the PR can be improved.

File: scripts/lockfile/lockfile.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/m3db/m3/src/x/lockfile"
+)
+
+func exitWithUsage() {
+	fmt.Printf("Usage: %s PATH SLEEP_SECONDS SHOULD_REMOVE_LOCK\n", path.Base(os.Args[0]))
```
Comment (richardartoul) [inline scripts/lockfile/lockfile.go]: nit: "Usage: %s <LOCK_FILE_PATH> <NUM_SECONDS_TO_SLEEP> <SHOULD_CLEANUP_LOCK>\n"

Comment (richardartoul) [reply scripts/lockfile/lockfile.go]: and then maybe one more line with a concrete example

File: src/x/lockfile/lockfile.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+type Lockfile struct {
+	file os.File
+}
+
+// Create creates the given file path if it doesn't exist and
+// obtains an exclusive lock on it. An error is returned if the lock
+// has been obtained by another process.
+func Create(path string) (*Lockfile, error) {
```
Comment (richardartoul) [inline src/x/lockfile/lockfile.go]: Could we call these `Lock` and `Unlock` or `Acquire` and `Release`

File: src/x/lockfile/lockfile.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0666)
```
Comment (richardartoul) [inline src/x/lockfile/lockfile.go]: what happens if some of the intermediary required directories don't exist? Will this still work or do you need a call to `os.MkDirAll()`

File: src/x/lockfile/lockfile.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	ft := &syscall.Flock_t{
+		Pid:  int32(os.Getpid()),
+		Type: syscall.F_UNLCK,
+	}
+
+	if err := syscall.FcntlFlock(lf.file.Fd(), syscall.F_SETLK, ft); err != nil {
+		return errors.Wrap(err, "failed releasing lock")
+	}
+
+	if err := os.Remove(lf.file.Name()); err != nil {
```
Comment (richardartoul) [inline src/x/lockfile/lockfile.go]: is it idiomatic to remove and then close?

File: src/x/lockfile/lockfile.go (hunk 3 above)
Comment (edkolev) [reply src/x/lockfile/lockfile.go]: the caller will have to call `MkDirAll()` or he won't be able to obtain the lock

File: src/x/lockfile/lockfile.go (hunk 4 above)
Comment (edkolev) [reply src/x/lockfile/lockfile.go]: I did some quick research but I didn't see any example of first remove then close -- I modified the code to close, then remove. I was thinking that there might be a race condition:
- process A closes the file handle
- process B creates the file and obtains the lock
- process A remove the files
- at this point B has a lock on a file which has been removed from the file system

But as I said, the code now closes, then removes, the above scenario doesn't seem likely. Let me know if you think we should dig deeper here and I'll try to reproduce the above scenario to check if it's a real concern.

File: src/x/lockfile/lockfile.go (hunk 2 above)
Comment (edkolev) [reply src/x/lockfile/lockfile.go]: renamed

**Comment by edkolev** [comment]
@richardartoul 

The PR is now updated with the feedback. I checked that `fcntl` is specified in POSIX, while `flock` isn't (which is the opposite to what I remembered). [posix link](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fcntl.html)

Should I squash the commits?

Thanks, and please let me know if I can further improve the PR.

File: src/x/lockfile/lockfile.go (hunk 4 above)
Comment (richardartoul) [reply src/x/lockfile/lockfile.go]: @edkolev If you delete a file that another process has a handle on, the file won't actually be deleted until that process releases the file

File: src/x/lockfile/lockfile.go (hunk 3 above)
Comment (richardartoul) [reply src/x/lockfile/lockfile.go]: Can we add it here so callers don't have to think about it? Maybe can even be an additional function:

```golang
func CreateAndAcquire(path string, newDirMode os.FileMode) (*LockFile, error) {
    if err := os.MkdirAll(paths.Dir(path), newDirMode); err != nil {
        return nil, err
    }

    return Acquire(path)
}
```

**Pull request merged**